From 89fee834756bbe6d9e060376f95188d66789b1f7 Mon Sep 17 00:00:00 2001
From: Sebastian Bazley <sebb@apache.org>
Date: Sun, 29 Jun 2014 17:44:58 +0000
Subject: [PATCH] JCS-133 RemoteUtils.getNamingURL does not handle IPv6 numeric
 addresses properly Fix could be improved; check if it helps the Jenkins build

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/jcs/trunk@1606543 13f79535-47bb-0310-9956-ffa450edef68
---
 .../apache/commons/jcs/auxiliary/remote/RemoteUtils.java   | 3 +++
 .../commons/jcs/auxiliary/remote/RemoteUtilsUnitTest.java  | 7 +++++++
 src/changes/changes.xml                                    | 3 +++
 3 files changed, 13 insertions(+)

diff --git a/commons-jcs-core/src/main/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtils.java b/commons-jcs-core/src/main/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtils.java
index 3bea93431..1962840ec 100644
--- a/commons-jcs-core/src/main/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtils.java
+++ b/commons-jcs-core/src/main/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtils.java
@@ -215,6 +215,9 @@ public ServerSocket createServerSocket( int port )
      */
     public static String getNamingURL(final String registryHost, final int registryPort, final String serviceName)
     {
+        if (registryHost.contains(":")) { // TODO improve this check? See also JCS-133
+            return "//[" + registryHost.replaceFirst("%", "%25") + "]:" + registryPort + "/" + serviceName;
+        }
         final String registryURL = "//" + registryHost + ":" + registryPort + "/" + serviceName;
         return registryURL;
     }
diff --git a/commons-jcs-core/src/test/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtilsUnitTest.java b/commons-jcs-core/src/test/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtilsUnitTest.java
index 7a557928c..3f4e42866 100644
--- a/commons-jcs-core/src/test/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtilsUnitTest.java
+++ b/commons-jcs-core/src/test/java/org/apache/commons/jcs/auxiliary/remote/RemoteUtilsUnitTest.java
@@ -41,4 +41,11 @@ public void testCreateRegistry()
         Registry registry = RemoteUtils.createRegistry( 1102 );
         assertNotNull("Registry should not be null", registry);
     }
+
+    public void testgetNamingURL()
+    {
+        assertEquals("//host:1/servicename", RemoteUtils.getNamingURL("host",1,"servicename"));
+        assertEquals("//127.0.0.1:2/servicename", RemoteUtils.getNamingURL("127.0.0.1",2,"servicename"));
+        assertEquals("//[0:0:0:0:0:0:0:1%251]:3/servicename", RemoteUtils.getNamingURL("0:0:0:0:0:0:0:1%1",3,"servicename"));
+    }
 }
diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 3e079f7d1..27a9e2d85 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -20,6 +20,9 @@
 	</properties>
 	<body>
 		<release version="2.0" date="unreleased" description="JDK 1.6 based major release">
+            <action issue="JCS-133" dev="seb" type="fix">
+                RemoteUtils.getNamingURL does not handle IPv6 numeric addresses properly
+            </action>
             <action issue="JCS-131" dev="tv" type="remove">
                 Remove KeyGeneratorUtil and AddressUtil as they are not used
             </action>
